Easy to fall into trap: Using a Property<ByteArray> in a custom gradle type

### Current Behavior

I tried to do the following:

```
interface MyWorkParams : WorkParameters {
   val byteArray: Property<ByteArray>
}
```

and got the following error:
```
Caused by: java.lang.ClassCastException: class [B cannot be cast to class [Ljava.lang.Object; ([B and [Ljava.lang.Object; are in module java.base of loader 'bootstrap')
```

### Expected Behavior

In Kotlin, the concept of primitives doesn't exist on the surface. And in the [gradle docs for custom types](https://docs.gradle.org/current/userguide/custom_gradle_types.html), searching for "array" or "primitive" or "bytearray" comes up with nothing. And the error message itself is very unfriendly to less experienced developers who have no idea what `[B` is. The following could all be helpful for this issue:
1. A better exception message. Catch the class cast exception, and throw a new exception with a more helpful message for Gradle users.
2. Clearer documentation. What should developers do when they want to use a type such as ByteArray in a custom gradle type?
3. Support ByteArray and similar types (arrays, primitives) directly. It seems like Gradle sees an Array and expects `[Ljava.lang.Object`. Is it possible it could be a bit more smart, and be prepared to see a `[B`?
4. If neccesarry, provide an ArrayProperty or ByteArrayProperty or PrimitveArrayProperty or PrimitiveProperty or whatever is neccesarry to support this. Though it would be ideal if it just worked inside of Property.

### Context (optional)

_No response_

### Steps to Reproduce

The code above should be sufficient.

### Gradle version

8.3

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/AbstractValueProcessor.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/AbstractValueProcessor.java
@@ -171,7 +171,7 @@ abstract class AbstractValueProcessor {
             return visitor.emptyArray(componentType);
         }
         if (componentType.isPrimitive()) {
-            throw cantSerializePrimitiveArray(valueClass);
+            return visitor.primitiveArray(value, valueClass);
         }
         return visitor.array(processArrayElements(value, length, visitor), componentType);
     }
@@ -205,7 +205,7 @@ abstract class AbstractValueProcessor {
         return visitor.javaSerialized(value, javaSerialized(value));
     }
 
-    private static byte[] javaSerialized(Object value) {
+    public static byte[] javaSerialized(Object value) {
         ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
         try (ObjectOutputStream oos = new ObjectOutputStream(outputStream)) {
             oos.writeObject(value);
@@ -265,6 +265,8 @@ abstract class AbstractValueProcessor {
         T emptyArray(Class<?> arrayType);
 
         T array(ImmutableList<T> elements, Class<?> arrayType);
+
+        T primitiveArray(Object value, Class<?> arrayType);
 
         T emptyList();
 

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/AbstractValueProcessor.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/AbstractValueProcessor.java
@@ -89,7 +89,7 @@ abstract class AbstractValueProcessor {
             return processMap((Map<?, ?>) value, visitor);
         }
         if (valueClass.isArray()) {
-            return processArray(valueClass, value, visitor);
+            return processArray(value, visitor);
         }
         if (value instanceof Attribute) {
             return visitor.attributeValue((Attribute<?>) value);
@@ -164,14 +164,14 @@ abstract class AbstractValueProcessor {
         return builder.build();
     }
 
-    private <T> T processArray(Class<?> valueClass, Object value, ValueVisitor<T> visitor) {
-        Class<?> componentType = valueClass.getComponentType();
+    private <T> T processArray(Object value, ValueVisitor<T> visitor) {
+        Class<?> componentType = value.getClass().getComponentType();
         int length = Array.getLength(value);
         if (length == 0) {
             return visitor.emptyArray(componentType);
         }
         if (componentType.isPrimitive()) {
-            return visitor.primitiveArray(value, valueClass);
+            return visitor.primitiveArray(value);
         }
         return visitor.array(processArrayElements(value, length, visitor), componentType);
     }
@@ -266,7 +266,7 @@ abstract class AbstractValueProcessor {
 
         T array(ImmutableList<T> elements, Class<?> arrayType);
 
-        T primitiveArray(Object value, Class<?> arrayType);
+        T primitiveArray(Object value);
 
         T emptyList();
 

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultIsolatableFactory.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultIsolatableFactory.java
@@ -157,8 +157,8 @@ public class DefaultIsolatableFactory extends AbstractValueProcessor implements 
         }
 
         @Override
-        public Isolatable<?> primitiveArray(Object value, Class<?> arrayType) {
-            return new IsolatedArrayOfPrimitive(value, arrayType);
+        public Isolatable<?> primitiveArray(Object value) {
+            return new IsolatedArrayOfPrimitive(value);
         }
 
         @Override

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultIsolatableFactory.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultIsolatableFactory.java
@@ -157,6 +157,11 @@ public class DefaultIsolatableFactory extends AbstractValueProcessor implements 
         }
 
         @Override
+        public Isolatable<?> primitiveArray(Object value, Class<?> arrayType) {
+            return new IsolatedArrayOfPrimitive(value, arrayType);
+        }
+
+        @Override
         public Isolatable<?> emptyList() {
             return IsolatedList.EMPTY;
         }

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultValueSnapshotter.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultValueSnapshotter.java
@@ -144,7 +144,11 @@ public class DefaultValueSnapshotter extends AbstractValueProcessor implements V
 
         @Override
         public ValueSnapshot javaSerialized(Object value, byte[] serializedValue) {
-            return new JavaSerializedValueSnapshot(classLoaderHasher.getClassLoaderHash(value.getClass().getClassLoader()), serializedValue);
+            return newJavaSerializedValueSnapshot(value.getClass(), serializedValue);
+        }
+
+        private JavaSerializedValueSnapshot newJavaSerializedValueSnapshot(Class<?> valueClass, byte[] serializedValue) {
+            return new JavaSerializedValueSnapshot(classLoaderHasher.getClassLoaderHash(valueClass.getClassLoader()), serializedValue);
         }
 
         @Override
@@ -155,6 +159,11 @@ public class DefaultValueSnapshotter extends AbstractValueProcessor implements V
         @Override
         public ValueSnapshot array(ImmutableList<ValueSnapshot> elements, Class<?> arrayType) {
             return new ArrayValueSnapshot(elements);
+        }
+
+        @Override
+        public ValueSnapshot primitiveArray(Object value, Class<?> arrayType) {
+            return newJavaSerializedValueSnapshot(arrayType, AbstractValueProcessor.javaSerialized(value));
         }
 
         @Override

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultValueSnapshotter.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/DefaultValueSnapshotter.java
@@ -144,11 +144,7 @@ public class DefaultValueSnapshotter extends AbstractValueProcessor implements V
 
         @Override
         public ValueSnapshot javaSerialized(Object value, byte[] serializedValue) {
-            return newJavaSerializedValueSnapshot(value.getClass(), serializedValue);
-        }
-
-        private JavaSerializedValueSnapshot newJavaSerializedValueSnapshot(Class<?> valueClass, byte[] serializedValue) {
-            return new JavaSerializedValueSnapshot(classLoaderHasher.getClassLoaderHash(valueClass.getClassLoader()), serializedValue);
+            return new JavaSerializedValueSnapshot(classLoaderHasher.getClassLoaderHash(value.getClass().getClassLoader()), serializedValue);
         }
 
         @Override
@@ -162,8 +158,8 @@ public class DefaultValueSnapshotter extends AbstractValueProcessor implements V
         }
 
         @Override
-        public ValueSnapshot primitiveArray(Object value, Class<?> arrayType) {
-            return newJavaSerializedValueSnapshot(arrayType, AbstractValueProcessor.javaSerialized(value));
+        public ValueSnapshot primitiveArray(Object value) {
+            return javaSerialized(value, AbstractValueProcessor.javaSerialized(value));
         }
 
         @Override

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
@@ -17,7 +17,6 @@
 package org.gradle.internal.snapshot.impl;
 
 import org.gradle.api.NonNullApi;
-import org.gradle.internal.hash.Hashable;
 import org.gradle.internal.hash.Hasher;
 import org.gradle.internal.isolation.Isolatable;
 import org.gradle.internal.snapshot.ValueSnapshot;
@@ -29,7 +28,7 @@ import static org.gradle.internal.snapshot.impl.AbstractValueProcessor.javaSeria
 import static org.gradle.internal.snapshot.impl.JavaSerializedValueSnapshot.javaDeserialized;
 
 @NonNullApi
-public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
+public class IsolatedArrayOfPrimitive implements Isolatable<Object> {
     private final PrimitiveType primitiveType;
     private final Object array;
 
@@ -61,28 +60,12 @@ public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
     }
 
     private UnsupportedOperationException unsupportedOperation() {
-        return new UnsupportedOperationException("This operation is not supported on primitive arrays of type: " + primitiveType.getArrayClass());
+        return new UnsupportedOperationException("This operation is not supported on primitive arrays of type: " + primitiveType.arrayType);
     }
 
     @Override
     public Object isolate() {
-        switch (primitiveType) {
-            case B:
-                return ((byte[]) array).clone();
-            case S:
-                return ((short[]) array).clone();
-            case I:
-                return ((int[]) array).clone();
-            case J:
-                return ((long[]) array).clone();
-            case F:
-                return ((float[]) array).clone();
-            case D:
-                return ((double[]) array).clone();
-            case C:
-                return ((char[]) array).clone();
-        }
-        throw new IllegalStateException();
+        return primitiveType.clone(array);
     }
 
     @Nullable
@@ -109,40 +92,67 @@ public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
             case B:
                 return bytes;
             default:
-                return javaDeserialized(primitiveType.getArrayClass(), bytes);
+                return javaDeserialized(primitiveType.arrayType, bytes);
         }
     }
 
     private enum PrimitiveType {
-        B,
-        S,
-        I,
-        J,
-        F,
-        D,
-        C;
+        B(byte[].class) {
+            @Override
+            public Object clone(Object array) {
+                return ((byte[]) array).clone();
+            }
+        },
+        S(short[].class) {
+            @Override
+            public Object clone(Object array) {
+                return ((short[]) array).clone();
+            }
+        },
+        I(int[].class) {
+            @Override
+            public Object clone(Object array) {
+                return ((int[]) array).clone();
+            }
+        },
+        J(long[].class) {
+            @Override
+            public Object clone(Object array) {
+                return ((long[]) array).clone();
+            }
+        },
+        F(float[].class) {
+            @Override
+            public Object clone(Object array) {
+                return ((float[]) array).clone();
+            }
+        },
+        D(double[].class) {
+            @Override
+            public Object clone(Object array) {
+                return ((double[]) array).clone();
+            }
+        },
+        C(char[].class) {
+            @Override
+            public Object clone(Object array) {
+                return ((char[]) array).clone();
+            }
+        };
+
+        public final Class<?> arrayType;
+
+        PrimitiveType(Class<?> arrayType) {
+            this.arrayType = arrayType;
+        }
+
+        public abstract Object clone(Object array);
 
         public static PrimitiveType of(Class<?> arrayType) {
-            if (byte[].class == arrayType) {
-                return B;
-            }
-            if (short[].class == arrayType) {
-                return S;
-            }
-            if (int[].class == arrayType) {
-                return I;
-            }
-            if (long[].class == arrayType) {
-                return J;
-            }
-            if (float[].class == arrayType) {
-                return F;
-            }
-            if (double[].class == arrayType) {
-                return D;
-            }
-            if (char[].class == arrayType) {
-                return C;
+            for (PrimitiveType primitiveType : values()) {
+                if (primitiveType.arrayType == arrayType) {
+                    return primitiveType;
+                }
             }
             throw new ValueSnapshottingException("Unsupported primitive array type: " + arrayType);
         }
@@ -151,26 +161,6 @@ public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
             PrimitiveType[] primitiveTypes = values();
             assert ordinal >= 0 && ordinal < primitiveTypes.length;
             return primitiveTypes[ordinal];
-        }
-
-        public Class<?> getArrayClass() {
-            switch (this) {
-                case B:
-                    return byte[].class;
-                case S:
-                    return short[].class;
-                case I:
-                    return int[].class;
-                case J:
-                    return long[].class;
-                case F:
-                    return float[].class;
-                case D:
-                    return double[].class;
-                case C:
-                    return char[].class;
-            }
-            throw new IllegalStateException();
         }
     }
 }

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
@@ -1,0 +1,168 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.snapshot.impl;
+
+import org.gradle.internal.hash.Hashable;
+import org.gradle.internal.hash.Hasher;
+import org.gradle.internal.isolation.Isolatable;
+import org.gradle.internal.snapshot.ValueSnapshot;
+import org.gradle.internal.snapshot.ValueSnapshottingException;
+
+import javax.annotation.Nullable;
+
+import static org.gradle.internal.snapshot.impl.AbstractValueProcessor.javaSerialized;
+import static org.gradle.internal.snapshot.impl.JavaSerializedValueSnapshot.javaDeserialized;
+
+public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
+    private final PrimitiveType primitiveType;
+    private final Object array;
+
+    public IsolatedArrayOfPrimitive(Object array, Class<?> arrayType) {
+        this(PrimitiveType.of(arrayType), array);
+    }
+
+    private IsolatedArrayOfPrimitive(PrimitiveType primitiveType, Object array) {
+        this.primitiveType = primitiveType;
+        this.array = array;
+    }
+
+    public static IsolatedArrayOfPrimitive decode(byte primitiveTypeCode, byte[] bytes) {
+        PrimitiveType primitiveType = PrimitiveType.values()[primitiveTypeCode];
+        return new IsolatedArrayOfPrimitive(
+            primitiveType,
+            fromByteArray(bytes, primitiveType)
+        );
+    }
+
+    @Override
+    public void appendToHasher(Hasher hasher) {
+        throw unsupportedOperation();
+    }
+
+    @Override
+    public ValueSnapshot asSnapshot() {
+        throw unsupportedOperation();
+    }
+
+    private UnsupportedOperationException unsupportedOperation() {
+        return new UnsupportedOperationException("This operation is not supported on primitive arrays of type: " + primitiveType.getArrayClass());
+    }
+
+    @Override
+    public Object isolate() {
+        switch (primitiveType) {
+            case B:
+                return ((byte[]) array).clone();
+            case S:
+                return ((short[]) array).clone();
+            case I:
+                return ((int[]) array).clone();
+            case J:
+                return ((long[]) array).clone();
+            case F:
+                return ((float[]) array).clone();
+            case D:
+                return ((double[]) array).clone();
+            case C:
+                return ((char[]) array).clone();
+        }
+        throw new IllegalStateException();
+    }
+
+    @Nullable
+    @Override
+    public <S> S coerce(Class<S> type) {
+        return null;
+    }
+
+    public byte getPrimitiveTypeCode() {
+        return (byte) primitiveType.ordinal();
+    }
+
+    public byte[] toByteArray() {
+        switch (primitiveType) {
+            case B:
+                return (byte[]) array;
+            default:
+                return javaSerialized(array);
+        }
+    }
+
+    private static Object fromByteArray(byte[] bytes, PrimitiveType primitiveType) {
+        switch (primitiveType) {
+            case B:
+                return bytes;
+            default:
+                return javaDeserialized(primitiveType.getArrayClass(), bytes);
+        }
+    }
+
+    private enum PrimitiveType {
+        B,
+        S,
+        I,
+        J,
+        F,
+        D,
+        C;
+
+        public static PrimitiveType of(Class<?> arrayType) {
+            if (byte[].class == arrayType) {
+                return B;
+            }
+            if (short[].class == arrayType) {
+                return S;
+            }
+            if (int[].class == arrayType) {
+                return I;
+            }
+            if (long[].class == arrayType) {
+                return J;
+            }
+            if (float[].class == arrayType) {
+                return F;
+            }
+            if (double[].class == arrayType) {
+                return D;
+            }
+            if (char[].class == arrayType) {
+                return C;
+            }
+            throw new ValueSnapshottingException("Unsupported primitive array type: " + arrayType);
+        }
+
+        public Class<?> getArrayClass() {
+            switch (this) {
+                case B:
+                    return byte[].class;
+                case S:
+                    return short[].class;
+                case I:
+                    return int[].class;
+                case J:
+                    return long[].class;
+                case F:
+                    return float[].class;
+                case D:
+                    return double[].class;
+                case C:
+                    return char[].class;
+            }
+            throw new IllegalStateException();
+        }
+    }
+}

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
@@ -16,6 +16,7 @@
 
 package org.gradle.internal.snapshot.impl;
 
+import org.gradle.api.NonNullApi;
 import org.gradle.internal.hash.Hashable;
 import org.gradle.internal.hash.Hasher;
 import org.gradle.internal.isolation.Isolatable;
@@ -27,6 +28,7 @@ import javax.annotation.Nullable;
 import static org.gradle.internal.snapshot.impl.AbstractValueProcessor.javaSerialized;
 import static org.gradle.internal.snapshot.impl.JavaSerializedValueSnapshot.javaDeserialized;
 
+@NonNullApi
 public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
     private final PrimitiveType primitiveType;
     private final Object array;

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/IsolatedArrayOfPrimitive.java
@@ -31,8 +31,8 @@ public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
     private final PrimitiveType primitiveType;
     private final Object array;
 
-    public IsolatedArrayOfPrimitive(Object array, Class<?> arrayType) {
-        this(PrimitiveType.of(arrayType), array);
+    public IsolatedArrayOfPrimitive(Object array) {
+        this(PrimitiveType.of(array.getClass()), array);
     }
 
     private IsolatedArrayOfPrimitive(PrimitiveType primitiveType, Object array) {
@@ -40,8 +40,8 @@ public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
         this.array = array;
     }
 
-    public static IsolatedArrayOfPrimitive decode(byte primitiveTypeCode, byte[] bytes) {
-        PrimitiveType primitiveType = PrimitiveType.values()[primitiveTypeCode];
+    public static IsolatedArrayOfPrimitive fromByteArray(byte primitiveTypeCode, byte[] bytes) {
+        PrimitiveType primitiveType = PrimitiveType.fromOrdinal(primitiveTypeCode);
         return new IsolatedArrayOfPrimitive(
             primitiveType,
             fromByteArray(bytes, primitiveType)
@@ -143,6 +143,12 @@ public class IsolatedArrayOfPrimitive implements Isolatable<Object>, Hashable {
                 return C;
             }
             throw new ValueSnapshottingException("Unsupported primitive array type: " + arrayType);
+        }
+
+        public static PrimitiveType fromOrdinal(byte ordinal) {
+            PrimitiveType[] primitiveTypes = values();
+            assert ordinal >= 0 && ordinal < primitiveTypes.length;
+            return primitiveTypes[ordinal];
         }
 
         public Class<?> getArrayClass() {

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/JavaSerializedValueSnapshot.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/internal/snapshot/impl/JavaSerializedValueSnapshot.java
@@ -81,6 +81,10 @@ public class JavaSerializedValueSnapshot implements ValueSnapshot {
     }
 
     protected Object populateClass(Class<?> originalClass) {
+        return javaDeserialized(originalClass, serializedValue);
+    }
+
+    public static Object javaDeserialized(Class<?> originalClass, byte[] serializedValue) {
         try {
             return new ClassLoaderObjectInputStream(new ByteArrayInputStream(serializedValue), originalClass.getClassLoader()).readObject();
         } catch (Exception e) {

--- a/platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParametersKotlinIntegrationTest.groovy
+++ b/platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParametersKotlinIntegrationTest.groovy
@@ -25,32 +25,55 @@ import spock.lang.Issue
 class WorkerExecutorParametersKotlinIntegrationTest extends AbstractIntegrationSpec {
 
     @Issue('https://github.com/gradle/gradle/issues/26596')
-    def "primitive #type array parameter in #isolationMode isolation fails with reasonable error message"() {
+    def "can provide primitive #type array parameter with #isolationMode isolation"() {
         given:
         withRunWorkTaskOf isolationMode, type
 
         when:
-        fails 'runWork'
+        succeeds 'runWork'
 
         then:
-        failureCauseContains "Cannot serialize primitive arrays of type $type[]"
+        outputContains expectedOutput
+
+        when:
+        succeeds 'runWork'
+
+        then:
+        outputDoesNotContain expectedOutput
+
+        and:
+        result.assertTaskSkipped ':runWork'
 
         where:
         [isolationMode, type] << [
             WorkerExecutorFixture.IsolationMode.values(),
-            ['byte', 'short', 'int', 'long', 'float', 'double']
+            ['byte', 'short', 'int', 'long', 'float', 'double', 'char']
         ].combinations()
+        expectedOutput = expectedOutputFor(type)
+    }
+
+    String expectedOutputFor(String type) {
+        switch (type) {
+            case 'char':
+                return 'charArrayOf(*)' // 42.toChar() == '*'
+            case 'float':
+            case 'double':
+                return "${type}ArrayOf(42.0)"
+            default:
+                return "${type}ArrayOf(42)"
+        }
     }
 
     private TestFile withRunWorkTaskOf(IsolationMode isolationMode, String primitiveType) {
         def kotlinType = primitiveType.toString().capitalize()
+        def kotlinArrayType = "${kotlinType}Array"
         buildKotlinFile << """
             import org.gradle.workers.WorkAction
             import org.gradle.workers.WorkParameters
             import org.gradle.workers.WorkerExecutor
 
             interface TestParameters : WorkParameters {
-                val array: Property<${kotlinType}Array>
+                val array: Property<$kotlinArrayType>
             }
 
             abstract class ParameterWorkAction : WorkAction<TestParameters> {
@@ -60,21 +83,31 @@ class WorkerExecutorParametersKotlinIntegrationTest extends AbstractIntegrationS
                 }
             }
 
+            @CacheableTask
             abstract class ParameterTask : DefaultTask() {
 
                 @get:Inject
                 abstract val workerExecutor: WorkerExecutor
 
+                @get:Input
+                abstract val inputArray: Property<$kotlinArrayType>
+
+                @get:OutputFile
+                abstract val outputFile: RegularFileProperty
+
                 @TaskAction
                 fun doWork() {
                     workerExecutor.${isolationMode.method}().submit(ParameterWorkAction::class) {
-                        array = ${primitiveType}ArrayOf(42.to${kotlinType}())
+                        array = inputArray
                     }
+                    outputFile.get().asFile.writeText("done")
                 }
             }
 
             tasks {
                 register<ParameterTask>("runWork") {
+                    inputArray = ${primitiveType}ArrayOf(42.to${kotlinType}())
+                    outputFile = layout.buildDirectory.file("receipt")
                 }
             }
         """

--- a/platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParametersKotlinIntegrationTest.groovy
+++ b/platforms/core-execution/workers/src/integTest/groovy/org/gradle/workers/internal/WorkerExecutorParametersKotlinIntegrationTest.groovy
@@ -73,6 +73,7 @@ class WorkerExecutorParametersKotlinIntegrationTest extends AbstractIntegrationS
             import org.gradle.workers.WorkerExecutor
 
             interface TestParameters : WorkParameters {
+                @get:Input
                 val array: Property<$kotlinArrayType>
             }
 

--- a/platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/DefaultActionExecutionSpecFactory.java
+++ b/platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/DefaultActionExecutionSpecFactory.java
@@ -45,22 +45,22 @@ public class DefaultActionExecutionSpecFactory implements ActionExecutionSpecFac
     @Override
     public <T extends WorkParameters> IsolatedParametersActionExecutionSpec<T> newIsolatedSpec(String displayName, Class<? extends WorkAction<T>> implementationClass, T params, WorkerRequirement workerRequirement, boolean usesInternalServices) {
         ClassLoaderStructure classLoaderStructure = workerRequirement instanceof IsolatedClassLoaderWorkerRequirement ? ((IsolatedClassLoaderWorkerRequirement) workerRequirement).getClassLoaderStructure() : null;
-        return new IsolatedParametersActionExecutionSpec<T>(implementationClass, displayName, implementationClass.getName(), isolatableFactory.isolate(params), classLoaderStructure, workerRequirement.getWorkerDirectory(), usesInternalServices);
+        return new IsolatedParametersActionExecutionSpec<>(implementationClass, displayName, implementationClass.getName(), isolatableFactory.isolate(params), classLoaderStructure, workerRequirement.getWorkerDirectory(), usesInternalServices);
     }
 
     @Override
     public <T extends WorkParameters> SimpleActionExecutionSpec<T> newSimpleSpec(IsolatedParametersActionExecutionSpec<T> spec) {
         T params = Cast.uncheckedCast(spec.getIsolatedParams().isolate());
-        return new SimpleActionExecutionSpec<T>(spec.getImplementationClass(), params, spec.isInternalServicesRequired());
+        return new SimpleActionExecutionSpec<>(spec.getImplementationClass(), params, spec.isInternalServicesRequired());
     }
 
     @Override
     public <T extends WorkParameters> SimpleActionExecutionSpec<T> newSimpleSpec(TransportableActionExecutionSpec spec) {
         T params = Cast.uncheckedCast(deserialize(spec.getSerializedParameters()).isolate());
-        return new SimpleActionExecutionSpec<T>(Cast.uncheckedCast(fromClassName(spec.getImplementationClassName())), params, spec.isInternalServicesRequired());
+        return new SimpleActionExecutionSpec<>(Cast.uncheckedCast(fromClassName(spec.getImplementationClassName())), params, spec.isInternalServicesRequired());
     }
 
-    Class<?> fromClassName(String className) {
+    static Class<?> fromClassName(String className) {
         try {
             return ClassLoaderUtils.classFromContextLoader(className);
         } catch (Exception e) {
@@ -70,8 +70,7 @@ public class DefaultActionExecutionSpecFactory implements ActionExecutionSpecFac
 
     private byte[] serialize(Isolatable<?> isolatable) {
         ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-        KryoBackedEncoder encoder = new KryoBackedEncoder(outputStream);
-        try {
+        try (KryoBackedEncoder encoder = new KryoBackedEncoder(outputStream)) {
             serializerRegistry.writeIsolatable(encoder, isolatable);
             encoder.flush();
         } catch (Exception e) {

--- a/platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/IsolatableSerializerRegistry.java
+++ b/platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/IsolatableSerializerRegistry.java
@@ -34,6 +34,7 @@ import org.gradle.internal.snapshot.impl.BooleanValueSnapshot;
 import org.gradle.internal.snapshot.impl.FileValueSnapshot;
 import org.gradle.internal.snapshot.impl.IntegerValueSnapshot;
 import org.gradle.internal.snapshot.impl.IsolatedArray;
+import org.gradle.internal.snapshot.impl.IsolatedArrayOfPrimitive;
 import org.gradle.internal.snapshot.impl.IsolatedEnumValueSnapshot;
 import org.gradle.internal.snapshot.impl.IsolatedImmutableManagedValue;
 import org.gradle.internal.snapshot.impl.IsolatedJavaSerializedValueSnapshot;
@@ -51,6 +52,7 @@ import org.gradle.internal.state.Managed;
 import org.gradle.internal.state.ManagedFactory;
 import org.gradle.internal.state.ManagedFactoryRegistry;
 
+import java.io.EOFException;
 import java.lang.reflect.Array;
 import java.util.Collection;
 import java.util.HashMap;
@@ -77,6 +79,8 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
     private static final byte ISOLATED_LIST = (byte) 14;
     private static final byte ISOLATED_SET = (byte) 15;
     private static final byte ISOLATED_PROPERTIES = (byte) 16;
+
+    private static final byte ISOLATED_ARRAY_OF_PRIMITIVE = (byte) 17;
 
     private static final byte ISOLATABLE_TYPE = (byte) 0;
     private static final byte ARRAY_TYPE = (byte) 1;
@@ -109,6 +113,7 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         isolatableSerializers.put(ISOLATED_LIST, new IsolatedListSerializer());
         isolatableSerializers.put(ISOLATED_SET, new IsolatedSetSerializer());
         isolatableSerializers.put(ISOLATED_PROPERTIES, new IsolatedPropertiesSerializer());
+        isolatableSerializers.put(ISOLATED_ARRAY_OF_PRIMITIVE, new IsolatedPrimitiveArraySerializer());
 
         for (IsolatableSerializer<?> serializer : isolatableSerializers.values()) {
             register(serializer.getIsolatableClass(), Cast.uncheckedCast(serializer));
@@ -174,7 +179,11 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
             writeIsolatable(encoder, (Isolatable<?>) state);
         } else if (state.getClass().isArray()) {
             encoder.writeByte(ARRAY_TYPE);
-            encoder.writeString(state.getClass().getComponentType().getName());
+            Class<?> componentType = state.getClass().getComponentType();
+            if (componentType.isPrimitive()) {
+                throw new IllegalStateException("Unsupported array type: " + state.getClass());
+            }
+            encoder.writeString(componentType.getName());
             Object[] array = (Object[]) state;
             int size = array.length;
             encoder.writeInt(size);
@@ -513,6 +522,31 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         @Override
         public Class<IsolatedProperties> getIsolatableClass() {
             return IsolatedProperties.class;
+        }
+    }
+
+    private class IsolatedPrimitiveArraySerializer implements IsolatableSerializer<IsolatedArrayOfPrimitive> {
+        @Override
+        public Class<IsolatedArrayOfPrimitive> getIsolatableClass() {
+            return IsolatedArrayOfPrimitive.class;
+        }
+
+        @Override
+        public void write(Encoder encoder, IsolatedArrayOfPrimitive value) throws Exception {
+            encoder.writeByte(ISOLATED_ARRAY_OF_PRIMITIVE);
+            encoder.writeByte(value.getPrimitiveTypeCode());
+            byte[] byteArray = value.toByteArray();
+            encoder.writeInt(byteArray.length);
+            encoder.writeBytes(byteArray);
+        }
+
+        @Override
+        public IsolatedArrayOfPrimitive read(Decoder decoder) throws EOFException, Exception {
+            byte primitiveTypeCode = decoder.readByte();
+            int length = decoder.readInt();
+            byte[] bytes = new byte[length];
+            decoder.readBytes(bytes);
+            return IsolatedArrayOfPrimitive.decode(primitiveTypeCode, bytes);
         }
     }
 

--- a/platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/IsolatableSerializerRegistry.java
+++ b/platforms/core-execution/workers/src/main/java/org/gradle/workers/internal/IsolatableSerializerRegistry.java
@@ -19,6 +19,7 @@ package org.gradle.workers.internal;
 import com.google.common.collect.ImmutableCollection;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableSet;
+import org.gradle.api.NonNullApi;
 import org.gradle.api.attributes.Attribute;
 import org.gradle.internal.Cast;
 import org.gradle.internal.hash.ClassLoaderHierarchyHasher;
@@ -52,7 +53,7 @@ import org.gradle.internal.state.Managed;
 import org.gradle.internal.state.ManagedFactory;
 import org.gradle.internal.state.ManagedFactoryRegistry;
 
-import java.io.EOFException;
+import java.io.IOException;
 import java.lang.reflect.Array;
 import java.util.Collection;
 import java.util.HashMap;
@@ -79,7 +80,6 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
     private static final byte ISOLATED_LIST = (byte) 14;
     private static final byte ISOLATED_SET = (byte) 15;
     private static final byte ISOLATED_PROPERTIES = (byte) 16;
-
     private static final byte ISOLATED_ARRAY_OF_PRIMITIVE = (byte) 17;
 
     private static final byte ISOLATABLE_TYPE = (byte) 0;
@@ -155,14 +155,7 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         } else if (stateType == ISOLATABLE_TYPE) {
             return readIsolatable(decoder);
         } else if (stateType == ARRAY_TYPE) {
-            String stateClassName = decoder.readString();
-            Class<?> stateClass = fromClassName(stateClassName);
-            int size = decoder.readInt();
-            Object state = Array.newInstance(stateClass, size);
-            for (int i = 0; i < size; i++) {
-                Array.set(state, i, readState(decoder));
-            }
-            return state;
+            return readArray(decoder);
         } else {
             String stateClassName = decoder.readString();
             Class<?> stateClass = fromClassName(stateClassName);
@@ -179,17 +172,7 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
             writeIsolatable(encoder, (Isolatable<?>) state);
         } else if (state.getClass().isArray()) {
             encoder.writeByte(ARRAY_TYPE);
-            Class<?> componentType = state.getClass().getComponentType();
-            if (componentType.isPrimitive()) {
-                throw new IllegalStateException("Unsupported array type: " + state.getClass());
-            }
-            encoder.writeString(componentType.getName());
-            Object[] array = (Object[]) state;
-            int size = array.length;
-            encoder.writeInt(size);
-            for (int i = 0; i < size; i++) {
-                writeState(encoder, array[i]);
-            }
+            writeArray(encoder, state);
         } else {
             encoder.writeByte(OTHER_TYPE);
             encoder.writeString(state.getClass().getName());
@@ -198,11 +181,35 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         }
     }
 
+    private Object readArray(Decoder decoder) throws Exception {
+        String componentTypeName = decoder.readString();
+        Class<?> componentType = fromClassName(componentTypeName);
+        int length = decoder.readInt();
+        Object state = Array.newInstance(componentType, length);
+        for (int i = 0; i < length; i++) {
+            Array.set(state, i, readState(decoder));
+        }
+        return state;
+    }
+
+    private void writeArray(Encoder encoder, Object state) throws Exception {
+        Class<?> componentType = state.getClass().getComponentType();
+        if (componentType.isPrimitive()) {
+            throw new IllegalArgumentException("Unsupported state array type: " + state.getClass());
+        }
+        encoder.writeString(componentType.getName());
+        Object[] array = (Object[]) state;
+        encoder.writeInt(array.length);
+        for (Object o : array) {
+            writeState(encoder, o);
+        }
+    }
+
     private interface IsolatableSerializer<T extends Isolatable<?>> extends Serializer<T> {
         Class<T> getIsolatableClass();
     }
 
-    private Class<?> fromClassName(String className) throws Exception {
+    private static Class<?> fromClassName(String className) {
         return classFromContextLoader(className);
     }
 
@@ -335,6 +342,7 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
 
             ManagedFactory factory = managedFactoryRegistry.lookup(factoryId);
             Managed managed = Cast.uncheckedCast(factory.fromState(publicClass, readState(decoder)));
+            assert managed != null;
             return new IsolatedImmutableManagedValue(managed, managedFactoryRegistry);
         }
 
@@ -350,8 +358,8 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
             encoder.writeByte(MANAGED_VALUE);
             encoder.writeInt(value.getFactoryId());
             encoder.writeString(value.getTargetType().getName());
-            Object state = value.getState();
-            writeIsolatable(encoder, (Isolatable<?>) state);
+            Isolatable<?> state = value.getState();
+            writeIsolatable(encoder, state);
         }
 
         @Override
@@ -389,7 +397,7 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         }
     }
 
-    private class IsolatedJavaSerializedValueSnapshotSerializer implements IsolatableSerializer<IsolatedJavaSerializedValueSnapshot> {
+    private static class IsolatedJavaSerializedValueSnapshotSerializer implements IsolatableSerializer<IsolatedJavaSerializedValueSnapshot> {
         @Override
         public void write(Encoder encoder, IsolatedJavaSerializedValueSnapshot value) throws Exception {
             encoder.writeByte(SERIALIZED_VALUE);
@@ -429,7 +437,7 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         }
 
         @Override
-        public NullValueSnapshot read(Decoder decoder) throws Exception {
+        public NullValueSnapshot read(Decoder decoder) {
             return NullValueSnapshot.INSTANCE;
         }
 
@@ -439,7 +447,7 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         }
     }
 
-    public class IsolatedEnumValueSnapshotSerializer implements IsolatableSerializer<IsolatedEnumValueSnapshot> {
+    public static class IsolatedEnumValueSnapshotSerializer implements IsolatableSerializer<IsolatedEnumValueSnapshot> {
         @Override
         public void write(Encoder encoder, IsolatedEnumValueSnapshot value) throws Exception {
             encoder.writeByte(ENUM_VALUE);
@@ -525,28 +533,37 @@ public class IsolatableSerializerRegistry extends DefaultSerializerRegistry {
         }
     }
 
-    private class IsolatedPrimitiveArraySerializer implements IsolatableSerializer<IsolatedArrayOfPrimitive> {
+    @NonNullApi
+    private static class IsolatedPrimitiveArraySerializer implements IsolatableSerializer<IsolatedArrayOfPrimitive> {
         @Override
         public Class<IsolatedArrayOfPrimitive> getIsolatableClass() {
             return IsolatedArrayOfPrimitive.class;
         }
 
         @Override
-        public void write(Encoder encoder, IsolatedArrayOfPrimitive value) throws Exception {
-            encoder.writeByte(ISOLATED_ARRAY_OF_PRIMITIVE);
-            encoder.writeByte(value.getPrimitiveTypeCode());
-            byte[] byteArray = value.toByteArray();
-            encoder.writeInt(byteArray.length);
-            encoder.writeBytes(byteArray);
+        public IsolatedArrayOfPrimitive read(Decoder decoder) throws Exception {
+            byte primitiveTypeCode = decoder.readByte();
+            byte[] bytes = readLengthPrefixedByteArray(decoder);
+            return IsolatedArrayOfPrimitive.fromByteArray(primitiveTypeCode, bytes);
         }
 
         @Override
-        public IsolatedArrayOfPrimitive read(Decoder decoder) throws EOFException, Exception {
-            byte primitiveTypeCode = decoder.readByte();
+        public void write(Encoder encoder, IsolatedArrayOfPrimitive value) throws Exception {
+            encoder.writeByte(ISOLATED_ARRAY_OF_PRIMITIVE);
+            encoder.writeByte(value.getPrimitiveTypeCode());
+            writeLengthPrefixedByteArray(encoder, value.toByteArray());
+        }
+
+        private static byte[] readLengthPrefixedByteArray(Decoder decoder) throws IOException {
             int length = decoder.readInt();
             byte[] bytes = new byte[length];
             decoder.readBytes(bytes);
-            return IsolatedArrayOfPrimitive.decode(primitiveTypeCode, bytes);
+            return bytes;
+        }
+
+        private static void writeLengthPrefixedByteArray(Encoder encoder, byte[] bytes) throws IOException {
+            encoder.writeInt(bytes.length);
+            encoder.writeBytes(bytes);
         }
     }
 

